Implement Find Again (F3), fixes #995

devdanzin opened the pull request:
Basic implementation for the Find Again feature suggested in #995: an `Editor.find_again() `  method that uses the  'find' parts of `find_replace()`. Testing is a copy-and-adapt deal right now. 

Refactoring the duplicated code into two helper methods (one for empty find string, one for present string) could make it more elegant. But that gets in the way if we want Find Again behavior to diverge from Find and Replace.

Currently, an empty find string will display a modal like for find and replace. We could mention `Ctrl+F` to provide a string (since the user didn't press that shortcut ). Or we could have F3 never pop up a modal, just a status bar message.

Open to and welcoming any suggestions, criticisms, alternative designs and requests.

dybber reviewed (approved):
Looks good, and works as expected. I can approve this.

devdanzin commented:
This has a problem: F3 is already used to "Flash your code onto the micro:bit" in that mode. Sorry about the oversight. Any suggestions for resolving this conflict?

tjguk commented:
@devdanzin I'd be interested in seeing this PR extended to include Shift-F3 which is a common key combo for reverse-search. If you had the opportunity, feel free. If you can't get to it, I'l merge this anyway as it is within the next day or so. Thanks :)

I've already pushed the question of the conflict with m:b mode to issue #1219 as I don't think it should stop this going ahead.

devdanzin commented:
Adding Shift+F3 done:
* added another shortcut;
* connect it to `find_again_backward`;
* made `find_again_backward` pass `forward=False` to `highlight_text`;
* added a workaround because `findFirst(forward=False)` doesn't  work, needs explicit line and index numbers.

I'd like to revisit this tomorrow, code is rather unpolished and tests need to check more details.

It seems we currently do a Find as if the input was a regex. If this isn't on purpose (I'll check history), it's a simple fix.

One of the changes brings a comment cheatsheet for `QsciScintilla.findFirst()` from `EditorPane.find_next_match`. It seems we don't use that method anywhere. We call `findFirst` with its opaque flags from `highlight_text` and `replace_text` instead. It might make sense to refactor this to at least always document `findFirst()` positional parameters.

tjguk commented:
Thanks, @devdanzin -- I hope to find time to review this tomorrow
